src/go/cmd/http-relay-server/server/broker.go

```diff
@@ -250,18 +256,30 @@ func (r *broker) SendResponse(resp *pb.HttpResponse) error {
 		brokerResponses.WithLabelValues("server_response", "invalid", backendName).Inc()
 		return fmt.Errorf("Duplicate or invalid request ID %s", id)
 	}
+	// hold `sendMutex` throughout the function to ensure that `responseStream` is not closed
+	// while we are writing to it.
+	pr.sendMutex.Lock()
```

Comment:
We need to be sure here that we aren't deadlocking because this is a potentially blocking operation while holding the global lock. Can you add a comment that explains why this is safe? And also for the other sendMutex.Lock() below.

Worst case is that one goroutine holds the global mutex and the other holds the sendMutex, and they can't be interrupted.

Reply:
Yeah, it could still deadlock the global mutex if `ReapActiveRequests` is blocked while trying to get `sendMutex` lock.

Reply:
Moving `close(markReap)` outside the critical section should avoid any deadlocks now.